### PR TITLE
fix: #34711 ensure distDir inside project folder

### LINKED ISSUE #34711: EACCES: permission denied, mkdir '../../dist'
### Verify canary release

- [X] I verified that the issue exists in Next.js canary release

### Provide environment information

```
$  npx --no-install next info
/bin/sh: 1: yarn: not found

    Operating System:
      Platform: linux
      Arch: x64
      Version: #1 SMP Wed Aug 25 23:20:18 UTC 2021
    Binaries:
      Node: 16.13.1
      npm: 8.5.1
      Yarn: N/A
      pnpm: 6.31.0
    Relevant packages:
      next: 12.1.0
      react: 17.0.2
      react-dom: 17.0.2
```

### What browser are you using? (if relevant)

Chrome

### How are you deploying your application? (if relevant)

nx.dev

### Describe the Bug

when run command `npx nx run blog:serve` I get the next error:

```
> Executing task: npx nx serve blog <


> nx run blog:serve

EACCES: permission denied, mkdir '../../dist'

 ———————————————————————

 >  NX   Running target "blog:serve" failed

   Failed tasks:
   
   - blog:serve
   
   Hint: run the command with --verbose for more details.

The terminal process "/usr/bin/bash '-c', 'npx nx serve blog'" terminated with exit code: 1.
```

### Expected Behavior

when run command `npx nx run blog:serve` to not have any error as it does with next: **12.0.7**.

### To Reproduce

```
git clone https://github.com/sincovschi/sincovschi-website.git
cd sincovschi-website
npm install
npm install next@latest
npx nx run blog:serve
```

### PR DESCRIPTION
**Pull request by sincovschi**
Fixes https://github.com/vercel/next.js/issues/34711

### CODE REVIEW

File: packages/next/server/dev/hot-reloader.ts [hunk 1]
```diff
@@ -147,6 +147,7 @@ export default class HotReloader {
   private buildId: string
   private middlewares: any[]
   private pagesDir: string
+  private distDir: string
   private webpackHotMiddleware?: WebpackHotMiddleware
   private config: NextConfigComplete
```
Comment (sincovschi) [inline packages/next/server/dev/hot-reloader.ts:152]:
```ts
console.log(this.config.distDir)
// '../../dist/apps/blog/.next'
```

File: packages/next/server/dev/next-dev-server.ts [hunk 2]
```diff
@@ -380,6 +380,7 @@ export default class DevServer extends Server {
 
     this.hotReloader = new HotReloader(this.dir, {
       pagesDir: this.pagesDir!,
+      distDir: this.distDir,
```
Comment (sincovschi) [inline packages/next/server/dev/next-dev-server.ts:383]:
```ts
this.distDir // here distDir is the absolute path
```

File: packages/next/server/dev/hot-reloader.ts [hunk 3]
```diff
@@ -435,15 +439,15 @@ export default class HotReloader {
     startSpan.stop() // Stop immediately to create an artificial parent span
 
     await this.clean(startSpan)
+
     // Ensure distDir exists before writing package.json
-    await fs.mkdir(this.config.distDir, { recursive: true })
+    await fs.mkdir(this.distDir, { recursive: true })
```
Comment (sincovschi) [inline packages/next/server/dev/hot-reloader.ts:444]: here `fs.mkdir` was trying to create a file from the root folder to a relative path that starts with `../../` which goes outside of the project.
therefore, I've replaced with `distDir` from the place of the Instance creation, where `distDir` is an absolute path and the issue dissappeared.

**Review by sincovschi — COMMENTED** [review]
Rechecked and left guiding comments.

**Review by timneutkens — APPROVED** [review]
Looks good 👍

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| buildDuration | 18.9s | 19.2s | ⚠️ +349ms |
| buildDurationCached | 7.4s | 7.7s | ⚠️ +291ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +280 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.676 | 3.593 | -0.08 |
| / avg req/sec | 680.15 | 695.87 | +15.72 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.807 | 1.755 | -0.05 |
| /error-in-render avg req/sec | 1383.57 | 1424.22 | +40.65 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.8 kB | 27.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 5.05 kB | 5.05 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| buildDuration | 23.1s | 22.4s | -679ms |
| buildDurationCached | 7.5s | 7.5s | -49ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +280 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.682 | 3.673 | -0.01 |
| / avg req/sec | 678.96 | 680.69 | +1.73 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.686 | 1.698 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1482.36 | 1472.5 | ⚠️ -9.86 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.1 kB | 42.1 kB | ✓ |
| main-HASH.js gzip | 27.8 kB | 27.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| image-HASH.js gzip | 5.08 kB | 5.08 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.28 kB | 2.28 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 534 B | 534 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: dbf1bcc56e676525a9c1078b37c2f026c05293d8

**Review by timneutkens — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| buildDuration | 14.9s | 14.8s | -62ms |
| buildDurationCached | 6s | 6s | -1ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +300 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.826 | 2.73 | -0.1 |
| / avg req/sec | 884.64 | 915.7 | +31.06 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.142 | 1.147 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 2188.4 | 2179.03 | ⚠️ -9.37 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 27.8 kB | 27.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.5 kB | 71.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 326 B | 326 B | ✓ |
| dynamic-HASH.js gzip | 2.57 kB | 2.57 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 919 B | 919 B | ✓ |
| image-HASH.js gzip | 5.05 kB | 5.05 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.26 kB | 2.26 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 530 B | 530 B | ✓ |
| link.html gzip | 544 B | 544 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| buildDuration | 18.4s | 18.3s | -66ms |
| buildDurationCached | 6s | 5.9s | -48ms |
| nodeModulesSize | 367 MB | 367 MB | ⚠️ +300 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.757 | 2.734 | -0.02 |
| / avg req/sec | 906.87 | 914.54 | +7.67 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.141 | 1.178 | ⚠️ +0.04 |
| /error-in-render avg req/sec | 2191.84 | 2122.28 | ⚠️ -69.56 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.1 kB | 42.1 kB | ✓ |
| main-HASH.js gzip | 27.8 kB | 27.8 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 71.6 kB | 71.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.56 kB | 2.56 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| image-HASH.js gzip | 5.08 kB | 5.08 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 2.28 kB | 2.28 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 14.7 kB | 14.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | sincovschi/next.js canary | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 547 B | ✓ |
| withRouter.html gzip | 527 B | 527 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

</details>
Commit: acf7b228d76708d7fad3986c4384a2357c549677

**Comment by svkale** [comment]
@ijjk @shuding @huozhi Please look up after this.
My development with Nx is stuck......

**Comment by timneutkens** [comment]
@svkale not sure what you're asking for, it was already merged and released on canary.

**Comment by svkale** [comment]
Sorry @timneutkens. I must be misunderstood. I am still getting the error for next 12.1.0. Do I need to install next '[12.1.1-canary.4](https://www.npmjs.com/package/next/v/12.1.1-canary.4)'

**Comment by balazsorban44** [comment]
Please open a bug report with a reproduction, so we can investigate :+1:

**Pull request merged**